pkg/replica/hash.go

```diff
@@ -0,0 +1,340 @@
+	SilentlyCorrupted bool
+}
+
+type SnapshotHashTask struct {
+	sync.Mutex
+
+	Ctx        context.Context
+	CancelFunc context.CancelFunc
+
+	SnapshotName string
+	Rehash       bool
+
+	file *os.File
+
+	SnapshotHashStatus
+}
+
+type SnapshotXattrHashInfo struct {
+	Method       string `json:"method"`
+	Checksum     string `json:"checksum"`
+	ModTime      string `json:"modTime"`
+	LastHashedAt string `json:"lastHashedAt"`
+}
+
+func NewSnapshotHashTask(ctx context.Context, cancel context.CancelFunc, snapshotName string, rehash bool) *SnapshotHashTask {
+	return &SnapshotHashTask{
+		Ctx:          ctx,
+		CancelFunc:   cancel,
+		SnapshotName: snapshotName,
+		Rehash:       rehash,
+
+		SnapshotHashStatus: SnapshotHashStatus{
+			State: ProgressStateInProgress,
+		},
+	}
+}
+
+func (t *SnapshotHashTask) LockFile() (fileLock *flock.Flock, err error) {
+	defer func() {
+		if err != nil && fileLock != nil && fileLock.Path() != "" {
+			if err := os.RemoveAll(fileLock.Path()); err != nil {
+				logrus.Warnf("failed to remove lock file %v since %v", fileLock.Path(), err)
+			}
+		}
+	}()
+
+	hostName := os.Getenv("HOSTNAME")
+	if hostName == "" {
+		return nil, fmt.Errorf("env HOSTNAME is missing")
+	}
+
+	err = os.MkdirAll(fileLockDirectory, 0755)
+	if err != nil {
+		return nil, err
+	}
+
+	fileLock = flock.New(filepath.Join(fileLockDirectory, hashLockFileName))
+
+	// Blocking lock
+	err = fileLock.Lock()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch the file lock for hashing snapshot %v", t.SnapshotName)
+	}
+
+	return fileLock, nil
+}
+
+func (t *SnapshotHashTask) UnlockFile(fileLock *flock.Flock) {
+	fileLock.Unlock()
+}
+
+func (t *SnapshotHashTask) Execute() {
+	var err error
+	var checksum string
+
+	defer func() {
+		t.StatusLock.Lock()
+		defer t.StatusLock.Unlock()
+		t.Checksum = checksum
+		if err != nil {
+			logrus.Errorf("failed to hash snapshot %v since %v", t.SnapshotName, err)
+			t.State = ProgressStateError
+			t.Error = err.Error()
+		} else {
+			t.State = ProgressStateComplete
+		}
+	}()
+
+	// Each node can have only snapshot hashing task at the same time.
+	// When the snapshot hashing task is initiated, the task will try to fetch the
+	// file lock (${fileLockDirectory}/hash), and other tasks need to wait for the lock.
+	// The file is unlocked after the task is completed.
+	fileLock, err := t.LockFile()
+	if err != nil {
+		return
+	}
+	defer t.UnlockFile(fileLock)
+
+	modTime, err := GetSnapshotModTime(t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	requireRehash := true
+	if !t.Rehash {
+		requireRehash, checksum = t.isRehashRequired(modTime)
+		if !requireRehash {
+			return
+		}
+	}
+
+	logrus.Infof("Starting hashing snapshot %v", t.SnapshotName)
+
+	lastHashedAt := time.Now().UTC().Format(time.RFC3339)
+	checksum, err = hashSnapshot(t.Ctx, t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	if t.isSnapshotSilentlyCorrupted(checksum) {
+		t.StatusLock.Lock()
+		t.SilentlyCorrupted = true
```

Comment:
Why do you choose to use a separate flag instead of directly marking the state as `errror` when the mismatching is detected?

Reply:
The `error` records the reason why the checksum cannot be successfully got.  If the silently corruption happens, the checksum still can be got.

Reply:
OK, got it. Then the next questions are, can we add one more state instead? Comparing this, which one is better?

Reply:
Good idea. Then, the slient corruption can be known quickly without the recalculation.

Reply:
Just for confirmation, you still prefer to use a separate field `SilentlyCorrupted` rather than introducing a new state for this case, right?